lib/reactor/builder.ex

```diff
@@ -209,127 +167,59 @@ defmodule Reactor.Builder do
+  def return!(reactor, name) do
+    case return(reactor, name) do
+      {:ok, reactor} -> reactor
+      {:error, reason} -> raise reason
+    end
   end
 
-  defp assert_is_step_impl({impl, opts}) when is_list(opts), do: assert_is_step_impl(impl)
+  @doc """
+  Compose another Reactor inside this one.
 
-  defp assert_is_step_impl(impl) when is_atom(impl) do
-    if Spark.implements_behaviour?(impl, Step) do
-      :ok
-    else
-      {:error,
-       ArgumentError.exception(
-         "Module `#{inspect(impl)}` does not implement the `Reactor.Step` behaviour."
-       )}
-    end
-  end
+  Whenever possible this function will extract the steps from inner Reactor and
+  place them inside the parent Reactor.  In order to achieve this the composer
+  will rename the steps to ensure that there are no conflicts.
 
-  defp build_argument_transform_steps(arguments, step_name) do
-    arguments
-    |> Enum.reduce_while({:ok, [], []}, fn
-      argument, {:ok, arguments, steps}
-      when is_from_input(argument) and has_transform(argument) ->
-        transform_step_name = {:__reactor__, :transform, argument.name, step_name}
-
-        step =
-          build_transform_step(
-            {:input, argument.source.name},
-            transform_step_name,
-            argument.transform
-          )
-
-        argument = %Argument{
-          name: argument.name,
-          source: %Template.Result{name: transform_step_name}
-        }
-
-        {:cont, {:ok, [argument | arguments], [%{step | transform: nil} | steps]}}
-
-      argument, {:ok, arguments, steps}
-      when is_from_result(argument) and has_transform(argument) ->
-        transform_step_name = {:__reactor__, :transform, argument.name, step_name}
-
-        step =
-          build_transform_step(
-            argument.source.name,
-            transform_step_name,
-            argument.transform
-          )
-
-        argument = %Argument{
-          name: argument.name,
-          source: %Template.Result{name: transform_step_name}
-        }
-
-        {:cont, {:ok, [argument | arguments], [%{step | transform: nil} | steps]}}
-
-      argument, {:ok, arguments, steps} when is_from_input(argument) ->
-        argument = %{argument | source: %Template.Result{name: {:input, argument.source.name}}}
-
-        {:cont, {:ok, [argument | arguments], steps}}
-
-      argument, {:ok, arguments, steps} when is_from_result(argument) ->
-        {:cont, {:ok, [argument | arguments], steps}}
-    end)
-  end
+  If you're attempting to create a recursive Reactor (ie compose a Reactor
+  within itself) then this will be detected and runtime composition will be used
+  instead.  See `Reactor.Step.Compose` for more details.
+  """
+  @spec compose(Reactor.t(), atom, Reactor.t() | module, [step_argument]) ::
+          {:ok, Reactor.t()} | {:error, any}
+  def compose(reactor, _name, _inner_reactor, _arguments) when not is_reactor(reactor),
```

Comment:
may eventually need an option here to explicitly prevent the fancy composition.

Reply:
I'm not 100% certain of that. If you want to do runtime composition you just do `Builder.add_step(:foo, {Step.Compose, reactor: inner_reactor, ...`